### PR TITLE
Andrey Rohau / Create test full response

### PR DESCRIPTION
**Pull request by andreyrogov8**

### CODE REVIEW

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 1]
```diff
@@ -157,48 +163,118 @@ private void CreateTestQuestionsForGrammar(List<Guid> questionIds, Test test)
...
+
+                return testQuestions;
+            }
+
+            private async Task<IQueryable<Question>> GetQuestionsByAuditionIdAsync(IEnumerable<TestQuestion> testQuestions, Guid? auditionId)
+            {
+                var questions = new List<Question>();
+                foreach (var testQuestion in testQuestions)
+                {
+                    var question = await _context.Questions.FirstOrDefaultAsync(x => x.Id == testQuestion.QuestionId).ConfigureAwait(false);
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:217]: database queries in a loop it's a bad practice. You should find solution with one query to the DB.

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 2]
```diff
@@ -157,48 +163,118 @@ private void CreateTestQuestionsForGrammar(List<Guid> questionIds, Test test)
...
+                        questions.Add(question);
+                    }
+                }
+                return questions.AsQueryable();
+            }
+            private async Task AddAnswerAsync(List<QuestionDto> questions)
+            {
+                foreach (var question in questions)
+                {
+                    question.Answers = await GetAnswerDtosAsync(question.Id).ConfigureAwait(false);
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:229]: database queries in a loop it's a bad practice. You should find solution with one query to the DB.

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 3]
```diff
@@ -157,48 +163,118 @@ private void CreateTestQuestionsForGrammar(List<Guid> questionIds, Test test)
                         QuestionId = question
                     };
                     _context.TestQuestions.Add(testQuestion);
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:165]: database queries in a loop it's a bad practice. You should find solution with one query to the DB.

**Review by DShcharbakou — CHANGES_REQUESTED** [review]

**Review by ristomin — DISMISSED** [review]

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 4]
```diff
@@ -156,49 +162,122 @@ private void CreateTestQuestionsForGrammar(List<Guid> questionIds, Test test)
...
+                var questionsList = await _context.Questions.AsNoTracking().ToListAsync().ConfigureAwait(false);
+                foreach (var testQuestion in testQuestions)
+                {
+                    var question = questionsList.FirstOrDefault(x => x.Id == testQuestion.QuestionId);
+                    if (question.AuditionId == auditionId)
+                    {
+                        fileteredQuestions.Add(question);
+                    }
+                }
+                return fileteredQuestions;
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:225]: In this method you can do something like this:
`
var testQuestionIds = testQuestions.Select(x => x.QuestionId);
return await _context.Questions.AsNoTracking().Where(x => x.AuditionId == auditionId && testQuestionIds.Contains(x.Id)).ToListAsync().ConfigureAwait(false);
`

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 5]
```diff
@@ -156,49 +162,122 @@ private void CreateTestQuestionsForGrammar(List<Guid> questionIds, Test test)
...
+            }
+            private async Task AddAnswersAsync(List<QuestionDto> questions)
+            {
+                var answerList = _context.Answers.AsNoTracking();
+                foreach (var question in questions)
+                {
+                    question.Answers = await GetAnswerDtosAsync(question.Id, answerList).ConfigureAwait(false);
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:232]: You should configurate mapping for List\<QuestionDto\> -> List\<Question\>

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 6]
```diff
@@ -156,49 +162,122 @@ private void CreateTestQuestionsForGrammar(List<Guid> questionIds, Test test)
...
+                }
+            }
+            private async Task<List<AnswerDto>> GetAnswerDtosAsync(Guid questionId, IQueryable<Answer> answerList)
+            {
+                var filteredAnswers = await answerList.Where(x => x.QuestionId == questionId)
+                    .ToListAsync().ConfigureAwait(false);
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:238]: You should do it in one query in AddAnswersAsync() line 229

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 7]
```diff
@@ -156,49 +162,122 @@ private void CreateTestQuestionsForGrammar(List<Guid> questionIds, Test test)
...
+
+                return answerDtos;
+            }
+            private List<QuestionDto> MapQuestions(IEnumerable<Question> questions)
+            {
+                var questionDtos = new List<QuestionDto>();
+
+                foreach (var question in questions)
+                {
+                    questionDtos.Add(_mapper.Map<QuestionDto>(question));
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:253]: You should configure List\<QuestionDto\> -> List\<Question\>

File: Elevel/Application/Features/TestCommands/CreateTestCommand.cs [hunk 8]
```diff
@@ -156,49 +162,122 @@ private void CreateTestQuestionsForGrammar(List<Guid> questionIds, Test test)
...
+                var audition = await _context.Auditions
+                    .ProjectTo<AuditionDto>(_mapper.ConfigurationProvider)
+                    .FirstOrDefaultAsync(x => x.Id == auditionId).ConfigureAwait(false);
+                audition.Questions = await GetQuestionDtosAsync(testId, testQuestions, auditionId).ConfigureAwait(false);
+                return audition;
+            }
+            private async Task<TopicDto> GetTopicAsync(Guid topicId)
+            {
+                var topic = await _context.Topics.ProjectTo<TopicDto>(_mapper.ConfigurationProvider).FirstOrDefaultAsync(x => x.Id == topicId).ConfigureAwait(false);
+                return topic;
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TestCommands/CreateTestCommand.cs:280]: just use 
`return await _context.Topics.ProjectTo<TopicDto>(_mapper.ConfigurationProvider).FirstOrDefaultAsync(x => x.Id == topicId).ConfigureAwait(false);`

**Review by DShcharbakou — CHANGES_REQUESTED** [review]

File: Elevel/Application/Profiles/TestProfile.cs [hunk 9]
```diff
@@ -10,13 +10,22 @@ public TestProfile()
...
+            CreateMap<Topic, CreateTestCommand.TopicDto>();
+            CreateMap<Audition, CreateTestCommand.AuditionDto>();
 
             CreateMap<Test, GetAllTestsQuery.TestDTO>();
 
             CreateMap<Test, UpdateTestCommand.Response>();
             CreateMap<UpdateTestCommand.Request, Test>();
 
             CreateMap<Test, GetTestByIdQuery.Response>();
+            CreateMap<Test, GetTestByIdQuery.Response>();
```
Comment (parys) [inline Elevel/Application/Profiles/TestProfile.cs:24]: duplicate 23-24

**Review by parys — APPROVED** [review]

**Review by ristomin — APPROVED** [review]

**Review by DShcharbakou — APPROVED** [review]

**Pull request merged**